File: 1658.minimum-operations-to-reduce-x-to-zero.py
#
# @lc app=leetcode id=1658 lang=python3
#
# [1658] Minimum Operations to Reduce X to Zero
#

# @lc code=start
class Solution:
    def minOperations(self, nums: list[int], x: int) -> int:
        total = sum(nums)
        left = 0
        ans = float('inf')
        for i in range(len(nums)):
            total -= nums[i]
            while total < x and left <= i:
                total += nums[left]
                left += 1
            #! only when we have a valid subarray, we can update the answer
            if total == x:
                ans = min(ans, len(nums) - (i - left + 1))
        return ans if ans != float('inf') else -1  
# @lc code=end

File: 1658.minimum-operations-to-reduce-x-to-zero.java
/*
 * @lc app=leetcode id=1658 lang=java
 *
 * [1658] Minimum Operations to Reduce X to Zero
 */

// @lc code=start
class Solution {
    public int minOperations(int[] nums, int x) {
        int left = 0, ans = Integer.MAX_VALUE;
        int sum = 0, cur = 0;
        for(int num : nums) sum += num;
        for(int right = 0; right < nums.length; right++){
            cur += nums[right];
            while(left <= right && sum - cur < x){
                cur -= nums[left++];
            }
            if(sum - cur == x){
                ans = Math.min(ans, nums.length - (right - left + 1));
            }

        }
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }
}
// @lc code=end

